Sparse_and_dense.ipynb

```diff
@@ -0,0 +1,320 @@
+    "    sparse.save_npz(f'data/{i}_sparse', sparse_mat)\n",
+    "    sparse.save_npz(f'data/{i}_sparse_un', sparse_mat, compressed=False)\n",
+    "    \n",
+    "    dense_size = os.path.getsize(f'data/{i}.npy') / 1000000\n",
+    "    sparse_size = os.path.getsize(f'data/{i}_sparse.npz') / 1000000\n",
+    "    sparse_un_size = os.path.getsize(f'data/{i}_sparse_un.npz') / 1000000\n",
+    "    print(f'dense: {dense_size :.4} MB\\nsparse: {sparse_size :.4} MB\\nsparse_uncompressed: {sparse_un_size :.4} MB\\n')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Memory-map reduces matrix-vector multiplication time for lower-density matrices\n",
+    "\n",
+    "Putting a normal load above a memmap load decreases memmap load time. However, loading memmap first does not speed up normal load times. Note, this only appears to be the case for very dense matrices. For best comparisons, we should always load a memmap matrix first in the sequence."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 110,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CCpGeAeGaD\n",
+      "density: 97.8 %\n",
+      "Normal: 0.000523 sec\n",
+      "MMAP: 0.00355 sec\n",
+      "MMAP2: 0.000354 sec\n",
+      "\n",
+      "DaGpBP\n",
+      "density: 21.0 %\n",
+      "Normal: 0.00228 sec\n",
+      "MMAP: 0.00106 sec\n",
+      "MMAP2: 0.00104 sec\n",
+      "\n",
+      "MFpGdCcSE\n",
+      "density: 14.1 %\n",
+      "Normal: 0.0234 sec\n",
+      "MMAP: 0.00703 sec\n",
+      "MMAP2: 0.00711 sec\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "for i in ['CCpGeAeGaD', 'DaGpBP', 'MFpGdCcSE']:\n",
+    "    print(i)\n",
+    "    mat = np.load(f'data/{i}.npy')\n",
+    "    \n",
+    "    density = 100 * (mat != 0).sum() / np.prod(mat.shape)\n",
+    "    print(f'density: {density :.3} %')\n",
+    "\n",
+    "#     Create a vector to multiply\n",
+    "    vector_size = mat.shape[0]\n",
+    "    vec = np.zeros((1, vector_size))\n",
+    "#     4 search nodes\n",
+    "    indices = np.random.randint(0, high=vector_size, size=4)\n",
+    "    vec[0, indices] = 1\n",
+    "    del mat\n",
+    "    \n",
+    "    t1 = time.time()\n",
+    "    matrix = np.load(f'data/{i}.npy', mmap_mode='r')\n",
+    "    output = vec @ matrix\n",
+    "    t2 = time.time()\n",
+    "    time_2 = t2 - t1\n",
+    "    del matrix\n",
+    "    del output\n",
+    "    \n",
+    "    t1 = time.time()\n",
+    "    matrix = np.load(f'data/{i}.npy', mmap_mode=None)\n",
+    "    output = vec @ matrix\n",
+    "    t2 = time.time()\n",
+    "    time_1 = t2 - t1\n",
+    "    del matrix\n",
+    "    del output\n",
+    "    \n",
+    "#     Second memory-map load time for comparison\n",
+    "    t1 = time.time()\n",
+    "    matrix = np.load(f'data/{i}.npy', mmap_mode='r')\n",
+    "    output = vec @ matrix\n",
```

Comment:
It'd be cool to see an additional comparison: Subset matrix for nonzero rows, then multiply. If memmap is very smart, it could be this efficient with the current method. If mmap is dumb, this could speed things up?

It may also affect the times of multiplying non-mmaped matrices.

Reply:
Not quite sure what you mean. If we subset the matrix after loading, then it would not have correct dimensions for the matrix-vector multiplication.

Reply:
You can subset the `matrix` and `vec` based on nonzero entries of `vec`... right?

Reply:
Oh I see what you mean: subset based on the nonzero rows, meaning rows corresponding to the search nodes. That makes sense, will try now.

Reply:
You were right about subsetting! It seems that so long as we create a new `np.ones` vector to multiply the subsetted matrix, then we get at least as quick load/multiplication times when compared to the standard memory map.